arwen/host/execution.go

```diff
@@ -532,3 +568,336 @@ func (host *vmHost) createETHCallInput() []byte {
+		vmOutput,
+		asyncCall.Destination,
+		callbackFunction,
+		executionError,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Callback omits for now any async call - TODO: take into consideration async calls generated from callbacks
+	callbackVMOutput, _, callBackErr := host.ExecuteOnDestContext(callbackCallInput)
+	err = host.processCallbackVMOutput(callbackVMOutput, callBackErr)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * savePendingAsyncCalls takes a list of pending async calls and save them to storage so the info will be available on callback
+ */
+func (host *vmHost) savePendingAsyncCalls(pendingAsyncMap *vmcommon.AsyncContextInfo) error {
+	storage := host.Storage()
+	runtime := host.Runtime()
+
+	asyncCallStorageKey, err := arwen.CustomStorageKey(host.Crypto(), arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(pendingAsyncMap)
+	if err != nil {
+		return err
+	}
+
+	_, err = storage.SetStorage(asyncCallStorageKey, data)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/**
+ * getPendingAsyncCalls returns only pending async calls from a list that can also contain resolved/rejected entries
+ */
+func (host *vmHost) getPendingAsyncCalls(asyncInfo *vmcommon.AsyncContextInfo) *vmcommon.AsyncContextInfo {
+	pendingMap := &vmcommon.AsyncContextInfo{
+		AsyncInitiator: vmcommon.AsyncInitiator{
+			CallerAddr: asyncInfo.CallerAddr,
+			ReturnData: asyncInfo.ReturnData,
+		},
+		AsyncContextMap: make(map[string]*vmcommon.AsyncContext),
+	}
+
+	for contextIdentifier, asyncContext := range asyncInfo.AsyncContextMap {
+		for _, asyncCall := range asyncContext.AsyncCalls {
+			if asyncCall.Status != vmcommon.AsyncCallPending {
+				continue
+			}
+			if pendingMap.AsyncContextMap[contextIdentifier] == nil {
+				pendingMap.AsyncContextMap[contextIdentifier] = &vmcommon.AsyncContext{
+					Callback: asyncContext.Callback,
+					AsyncCalls: make([]*vmcommon.AsyncGeneratedCall, 0),
+				}
+			}
+			pendingMap.AsyncContextMap[contextIdentifier].AsyncCalls = append(
+				pendingMap.AsyncContextMap[contextIdentifier].AsyncCalls,
+				asyncCall,
+			)
+		}
+	}
+
+	return pendingMap
+}
+
+/**
+ * processCallbackStack is triggered when a callback was received from another host through a transaction.
+ *  It will return an error if we receive a callback and we don't have it's associated data in the storage.
+ *  If the associated callback was found in the pending set, it will be removed - It should not be executed
+ *   again since it was executed in the callSCMethod step
+ */
+func (host *vmHost) processCallbackStack() error {
+	runtime := host.Runtime()
+	storage := host.Storage()
+
+	storageKey, err := arwen.CustomStorageKey(host.Crypto(), arwen.AsyncDataPrefix, runtime.GetOriginalTxHash())
+	if err != nil {
+		return err
+	}
+
+	buff := storage.GetStorage(storageKey)
+
+	asyncInfo := &vmcommon.AsyncContextInfo{}
+	err = json.Unmarshal(buff, &asyncInfo)
```

Comment:
might use marshalizer from vmHost

Reply:
I don't see `vmHost` having a marshalizer, am I not looking close enough?

Reply:
it could be added there. But it can be another PR - please add a TODO and a task in JIRA